website/blog/2023-07-14-Local-LLMs/index.mdx

```diff
@@ -0,0 +1,107 @@
+
+## Preparations
+
+### Clone FastChat
+
+FastChat provides OpenAI-compatible APIs for its supported models, so you can use FastChat as a local drop-in replacement for OpenAI APIs. However, it's code needs minor modification in order to function properly.
+
+```bash
+git clone https://github.com/lm-sys/FastChat.git
+cd FastChat
+```
+
+According to this [issue](https://github.com/lm-sys/FastChat/issues/1641), comment out all the line containing `finish_reason` in `fastchat/protocol/api_protocal.py` and `fastchat/protocol/openai_api_protocol.py` will fix the errors. This is what modified code looks like:
+
+```python
+class CompletionResponseChoice(BaseModel):
+    index: int
+    text: str
+    logprobs: Optional[int] = None
+    # finish_reason: Optional[Literal["stop", "length"]]
+
+class CompletionResponseStreamChoice(BaseModel):
+    index: int
+    text: str
+    logprobs: Optional[float] = None
+    # finish_reason: Optional[Literal["stop", "length"]] = None
+```
+
+### Download checkpoint
+
+ChatGLM-6B is an open bilingual language model based on General Language Model (GLM) framework, with 6.2 billion parameters. ChatGLM2-6B is its second-generation version.
+
+Before downloading from HuggingFace Hub, you need to have Git LFS [installed](https://docs.github.com/en/repositories/working-with-files/managing-large-files/installing-git-large-file-storage).
+
+```bash
+git clone https://huggingface.co/THUDM/chatglm2-6b
+```
+
+## Initiate server
+
+First, launch the controller
+
+```bash
+python -m fastchat.serve.controller
+```
+
+Then, launch the model worker(s)
+
+```bash
+python -m fastchat.serve.model_worker --model-path chatglm2-6b
+```
+
+Finally, launch the RESTful API server
+
+```bash
+python -m fastchat.serve.openai_api_server --host localhost --port 8000
+```
+
+## Interact with model using `oai.Completion`
+
+Now the models can be directly accessed through openai-python library as well as `flaml.oai.Completion`.
+
+```python
+from flaml import oai
+
+# create a text completion request
+response = oai.Completion.create(
+    config_list=[
+        {
+            "model": "chatglm2-6b",
+            "api_base": "http://localhost:8000/v1",
+            "api_type": "open_ai",
+            "api_key": "NULL", # just a placeholder
+        }
+    ],
+    prompt="Hi",
+)
+print(response)
+
+# create a chat completion request
+response = oai.Completion.create(
+    config_list=[
+        {
+            "model": "chatglm2-6b",
+            "api_base": "http://localhost:8000/v1",
+            "api_type": "open_ai",
+            "api_key": "NULL",
+        }
+    ],
+    messages=[{"role": "user", "content": "Hi"}]
+)
+print(response)
+```
+
+If you would like to switch to different models, download their checkpoints and specify model path when launching model worker(s). 
```

Comment:
Can one launch multiple model workers?

Reply:
I believe yes, specify multiple model path when launching workers will work. I will test it out.

Reply:
Thanks. Please document it if it works.

Reply:
Done.